# Pandas

In [1]:
# Importing numpy and pandas library
import numpy as np
import pandas as pd

### Pandas Series

One dimensional labeled array in Pandas series is capable of holding any type of data. Data alignment in Pandas is intrinsic. 

In [56]:
# Understanding what goes into a pandas series
pd.Series?

In [61]:
# Creating a pandas series from a list (similar to numpy.array)
my_list = [10,20,30]
print(my_list)
ds = pd.Series(data=my_list,index=[4,5,6])
type(ds) # printing the type of pandas data series

[10, 20, 30]


pandas.core.series.Series

In [62]:
# Printing the output
ds

4    10
5    20
6    30
dtype: int64

In [63]:
# Syntax for creating a pandas series is pd.Series(data,index)

# Creating a pandas series from numpy array
my_array = np.linspace(0,10,11)
ds = pd.Series(my_array)
print(my_array)
print(80*'-')
ds # Pandas series prints the index and values (unlike numpy array)

[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
--------------------------------------------------------------------------------


0      0.0
1      1.0
2      2.0
3      3.0
4      4.0
5      5.0
6      6.0
7      7.0
8      8.0
9      9.0
10    10.0
dtype: float64

In [64]:
# Playing around with the index
my_array = np.linspace(0,10,11)
my_index1 = np.linspace(10,20,11)
my_index2 = np.linspace(100,110,11)
ds1 = pd.Series(data=my_array,index=my_index1) 
ds2 = pd.Series(data=my_array,index=my_index2) # Two series with same vlaues but different index

In [65]:
ds1

10.0     0.0
11.0     1.0
12.0     2.0
13.0     3.0
14.0     4.0
15.0     5.0
16.0     6.0
17.0     7.0
18.0     8.0
19.0     9.0
20.0    10.0
dtype: float64

In [66]:
ds2

100.0     0.0
101.0     1.0
102.0     2.0
103.0     3.0
104.0     4.0
105.0     5.0
106.0     6.0
107.0     7.0
108.0     8.0
109.0     9.0
110.0    10.0
dtype: float64

In [73]:
# You can not compare the two dataframes (due to different indexes they are two different dataframes)
# For two pandas series to be comparable, they must be identical in all respects (index and values)
(ds1.values == ds2.values).any()


True

In [89]:
# Index can take any name (not only restricted to numbers)
my_array3 = np.array([1,2,3,4])
my_index3 = "A A C D".split()
ds3 = pd.Series(data=my_array3,index=my_index3)
ds3['A']


A    1
A    2
dtype: int32

In [91]:
# Index can take any name (not only restricted to numbers)
my_array3 = np.array([1,2,3,4])
my_index3 = ["A B C D".split()]
ds3 = pd.Series(data=my_array3,index=my_index3)

my_array4 = np.array([1,2,3,4])
my_index4 = ["B C D E".split()]
ds4 = pd.Series(data=my_array4,index=my_index4)


In [92]:
ds3

A    1
B    2
C    3
D    4
dtype: int32

In [93]:
ds4

B    1
C    2
D    3
E    4
dtype: int32

In [94]:
# When we try to add two series with different index values
ds3+ds4

A    NaN
B    3.0
C    5.0
D    7.0
E    NaN
dtype: float64

In [96]:
# What is NaN (arithmetic operations on nan)
np.nan

nan

In [100]:
#Resetting index
#ds.reset_index()
#type(ds.reset_index())
ds3.reset_index(drop=True)

0    1
1    2
2    3
3    4
dtype: int32

### Pandas Data Frames

Recall that a numpy matrix was created by stacking numpy arrays. Similarly, a Pandas dataframe is constructed by stacking pandas series together.

In [101]:
# Visualizing our data through a numpy matrix first
my_matrix = np.array([['Ronaldo', 2002, 'Brazil'],
                      ['Grosso', 2006, 'Italy'],
                      ['Iniesta', 2010, 'Spain'],
                      ['Gotze', 2014, 'Germany']])

my_matrix

array([['Ronaldo', '2002', 'Brazil'],
       ['Grosso', '2006', 'Italy'],
       ['Iniesta', '2010', 'Spain'],
       ['Gotze', '2014', 'Germany']], dtype='<U7')

In [102]:
pd.DataFrame?

In [243]:
# Creating a pandas dataframe from scratch
df = pd.DataFrame([['Andy', 87, 'M'],
                   ['Bill', 67, 'M'],
                   ['Catherine', 93, 'F'],
                   ['David', 95, 'M'],
                   ['Emma', 50, 'F']],
                  columns=['Name', 'Score', 'Gender'])

df

Name  Score Gender
0       Andy     87      M
1       Bill     67      M
2  Catherine     93      F
3      David     95      M
4       Emma     50      F

In [244]:
df.drop(columns=['Gender'],inplace=True)

In [245]:
df

Name  Score
0       Andy     87
1       Bill     67
2  Catherine     93
3      David     95
4       Emma     50

In [107]:
type(df)

pandas.core.frame.DataFrame

In [118]:
# Selecting a column
type(df[['Score']])

pandas.core.frame.DataFrame

In [113]:
# SQL way of selecting data (not the recommended way - not considered good a programming practice)
df.Score

0    87
1    67
2    93
3    95
4    50
Name: Score, dtype: int64

In [119]:
# Selecting two columns (spot the difference)
df[['Name','Score']]

Name  Score
0       Andy     87
1       Bill     67
2  Catherine     93
3      David     95
4       Emma     50

In [122]:
# Rearranging the column order
df = df[['Name','Gender','Score']] # not an inplace operation
df

Name Gender  Score
0       Andy      M     87
1       Bill      M     67
2  Catherine      F     93
3      David      M     95
4       Emma      F     50

In [123]:
# Location operator - syntax is .loc[row,column]
df.loc[0,["Name","Score"]]

Name     Andy
Score      87
Name: 0, dtype: object

In [125]:
# Modifying data using the Location operator
df.loc[0,"Score"] = 92
df

Name Gender  Score
0       Andy      M     92
1       Bill      M     67
2  Catherine      F     93
3      David      M     95
4       Emma      F     50

In [135]:
# Modifying the index and using the loc operator
df.index = "A B C D E".split()
df
df.loc["A","Gender"]
# Show error by changing A to 0

'M'

In [136]:
# Now using the integer location operator - syntax .iloc[row_integer,column_integer]
df.iloc[0,1]

'M'

In [137]:
# Showing column with iloc
df.iloc[:,1] # Rows and columns in iloc have to be integers (names will give an error)

A    M
B    M
C    F
D    M
E    F
Name: Gender, dtype: object

In [138]:
# When a series is compared to a constant (similar to numpy)
df[df['Score']>90]

Name Gender  Score
A       Andy      M     92
C  Catherine      F     93
D      David      M     95

### DataFrame Operations

In [154]:
# Creating a new column
df = pd.DataFrame([['stock1',50,1000],
                   ['stock2',75,900],
                   ['stock3',125,400]],columns=['stock_name','price','shares'])
df
df['market_cap'] = df['price']*df['shares']
df

stock_name  price  shares  market_cap
0     stock1     50    1000       50000
1     stock2     75     900       67500
2     stock3    125     400       50000

In [143]:
# Sum function can be called on the series (and on the dataframe)
df['market_cap'].prod()

168750000000000

In [155]:
# Create an empty dataframe column by broadcasting
df['pe_ratio'] = [1,2,3]
df

stock_name  price  shares  market_cap  pe_ratio
0     stock1     50    1000       50000         1
1     stock2     75     900       67500         2
2     stock3    125     400       50000         3

In [153]:
# Dropping the market_cap column
#df = df.drop(columns=['market_cap']) # inplace=False is default
#df
print(df.drop(columns=['market_cap'],inplace=True))

None


In [ ]:
# Dropping rows
df.drop(0) # not an inplace opearator

In [156]:
# Sorting the dataframe on outstanding shares
df.sort_values(by=['shares'])

stock_name  price  shares  market_cap  pe_ratio
2     stock3    125     400       50000         3
1     stock2     75     900       67500         2
0     stock1     50    1000       50000         1

In [157]:
# Setting index
df = df.set_index('stock_name')
df

price  shares  market_cap  pe_ratio
stock_name                                     
stock1         50    1000       50000         1
stock2         75     900       67500         2
stock3        125     400       50000         3

In [158]:
# Resetting index
df = df.reset_index()
df

stock_name  price  shares  market_cap  pe_ratio
0     stock1     50    1000       50000         1
1     stock2     75     900       67500         2
2     stock3    125     400       50000         3

In [168]:
# Creating a dataframe 
df = pd.DataFrame(np.linspace(0,8,9).reshape(3,3),columns=['A B C'.split()])
df

df.iloc[1,2] = np.nan
df.iloc[2,2] = np.nan

df

A    B    C
0  0.0  1.0  2.0
1  3.0  4.0  NaN
2  6.0  7.0  NaN

In [169]:
# Sum function 
df.sum(axis=0,skipna=True,min_count=2)


A     9.0
B    12.0
C     NaN
dtype: float64

### Data Analysis

In [170]:
# Loading panel data through pandas
df = pd.read_csv("stock_data.csv",sep = ',')
df.shape

(2960, 6)

In [175]:
# Viewing a sample of the dataframe (due to large size)
df.head(10)

permno     company_name             date total_returns    price  \
0   10001       ENERGY INC  12/31/2009 0:00      0.162621     10.3   
1   10001       ENERGY INC   1/31/2010 0:00     -0.018932    10.06   
2   10001       ENERGY INC   2/28/2010 0:00     -0.000656  10.0084   
3   10001       ENERGY INC   3/31/2010 0:00      0.020643    10.17   
4   10001       ENERGY INC   4/30/2010 0:00      0.124385    11.39   
5   10001       ENERGY INC   5/31/2010 0:00      0.004829     11.4   
6   10001       ENERGY INC   6/30/2010 0:00     -0.043421    10.86   
7   10001  GAS NATURAL INC   7/31/2010 0:00      0.083333    11.72   
8   10001  GAS NATURAL INC   8/31/2010 0:00     -0.111263   10.371   
9   10001  GAS NATURAL INC   9/30/2010 0:00       0.07656    11.12   

   share_outstanding  
0               4361  
1               4361  
2               4361  
3               4361  
4               6070  
5               6071  
6               6080  
7               6080  
8               6073  
9               6073

In [172]:
# Getting info on our dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2960 entries, 0 to 2959
Data columns (total 6 columns):
permno               2960 non-null int64
company_name         2960 non-null object
date                 2960 non-null object
total_returns        2933 non-null object
price                2920 non-null object
share_outstanding    2960 non-null int64
dtypes: int64(2), object(4)
memory usage: 138.8+ KB


In [173]:
# Unique permnos
df['permno'].nunique()

25

In [174]:
# Unique company names
df['company_name'].nunique()

35

In [176]:
# Observe how the date is initially loaded 
df['date'][0]

'12/31/2009 0:00'

In [179]:
# Pandas has a convenient hack for dates
df['date'] = pd.to_datetime(df['date'])
df['date'][0]

Timestamp('2009-12-31 00:00:00')

In [180]:
# Dataframe date format
df['date'].head()

0   2009-12-31
1   2010-01-31
2   2010-02-28
3   2010-03-31
4   2010-04-30
Name: date, dtype: datetime64[ns]

In [181]:
# Converting datetime from one format to another
df['date'].dt.strftime("%d-%m-%Y").head() 

0    31-12-2009
1    31-01-2010
2    28-02-2010
3    31-03-2010
4    30-04-2010
Name: date, dtype: object

In [182]:
# writing dates in pandas
pd.Timestamp(year=2013,month=12,day=7)

Timestamp('2013-12-07 00:00:00')

In [191]:
# Adding days to a date (or hours/minutes etc)
some_date = pd.Timestamp(year=2013,month=12,day=7)
print(some_date)
some_date + pd.Timedelta('3 days')
print(df['date'].head())
#print(df['date'].head()+pd.Timedelta("1 day"))


2013-12-07 00:00:00
0   2009-12-31
1   2010-01-31
2   2010-02-28
3   2010-03-31
4   2010-04-30
Name: date, dtype: datetime64[ns]


In [193]:
# Creating a Series of dates (a very cool trick!)
my_date = pd.to_datetime(pd.Series(['13 July 2013','5 May 2010','15 March 2020']))
print(my_date)
my_date + pd.tseries.offsets.MonthEnd(0)

0   2013-07-13
1   2010-05-05
2   2020-03-15
dtype: datetime64[ns]


0   2013-07-31
1   2010-05-31
2   2020-03-31
dtype: datetime64[ns]

In [197]:
# Using Pandas date range to generate dates
#pd.date_range?
#pd.date_range(start='01 Jan 2021', end='31 Jan 2021',periods=10)
pd.date_range(start='01 Jan 2021', periods=18, freq= 'M')

DatetimeIndex(['2021-01-31', '2021-02-28', '2021-03-31', '2021-04-30',
               '2021-05-31', '2021-06-30', '2021-07-31', '2021-08-31',
               '2021-09-30', '2021-10-31', '2021-11-30', '2021-12-31',
               '2022-01-31', '2022-02-28', '2022-03-31', '2022-04-30',
               '2022-05-31', '2022-06-30'],
              dtype='datetime64[ns]', freq='M')

In [204]:
# Loading the dataframe again
df.tail()

permno          company_name       date total_returns  price  \
2955   10137  ALLEGHENY ENERGY INC 2010-10-31     -0.053834   23.2   
2956   10137  ALLEGHENY ENERGY INC 2010-11-30           NaN  22.82   
2957   10137  ALLEGHENY ENERGY INC 2010-12-31             C      C   
2958   10137  ALLEGHENY ENERGY INC 2011-01-31      0.063531  25.78   
2959   10137  ALLEGHENY ENERGY INC 2011-02-28     -0.009073    NaN   

      share_outstanding  
2955             169615  
2956             169939  
2957             169939  
2958             169939  
2959             169974

In [198]:
# Checking the dataframe types (from info)
df.info() # observe that prices and returns are still objects!

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2960 entries, 0 to 2959
Data columns (total 6 columns):
permno               2960 non-null int64
company_name         2960 non-null object
date                 2960 non-null datetime64[ns]
total_returns        2933 non-null object
price                2920 non-null object
share_outstanding    2960 non-null int64
dtypes: datetime64[ns](1), int64(2), object(3)
memory usage: 138.8+ KB


In [205]:
# Converting price and returns to numbers
df['price'] = pd.to_numeric(df['price'],errors='coerce')
df['total_returns'] = pd.to_numeric(df['total_returns'],errors='coerce')

In [206]:
# Checking info again
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2960 entries, 0 to 2959
Data columns (total 6 columns):
permno               2960 non-null int64
company_name         2960 non-null object
date                 2960 non-null datetime64[ns]
total_returns        2929 non-null float64
price                2916 non-null float64
share_outstanding    2960 non-null int64
dtypes: datetime64[ns](1), float64(2), int64(2), object(1)
memory usage: 138.8+ KB


In [207]:
df.tail()

permno          company_name       date  total_returns  price  \
2955   10137  ALLEGHENY ENERGY INC 2010-10-31      -0.053834  23.20   
2956   10137  ALLEGHENY ENERGY INC 2010-11-30            NaN  22.82   
2957   10137  ALLEGHENY ENERGY INC 2010-12-31            NaN    NaN   
2958   10137  ALLEGHENY ENERGY INC 2011-01-31       0.063531  25.78   
2959   10137  ALLEGHENY ENERGY INC 2011-02-28      -0.009073    NaN   

      share_outstanding  
2955             169615  
2956             169939  
2957             169939  
2958             169939  
2959             169974

In [210]:
# Rows that have NaNs
df[df['total_returns'].isnull()] # Returns True in dataframe where values are NaN
# df.isnull().sum() gives count of NaN values

permno                      company_name       date  total_returns  \
1010   10051                        HANGER INC 2016-02-29            NaN   
1011   10051                        HANGER INC 2016-03-31            NaN   
1012   10051                        HANGER INC 2016-04-30            NaN   
1013   10051                        HANGER INC 2016-05-31            NaN   
1014   10051                        HANGER INC 2016-06-30            NaN   
1015   10051                        HANGER INC 2016-07-31            NaN   
1016   10051                        HANGER INC 2016-08-31            NaN   
1017   10051                        HANGER INC 2016-09-30            NaN   
1018   10051                        HANGER INC 2016-10-31            NaN   
1019   10051                        HANGER INC 2016-11-30            NaN   
1020   10051                        HANGER INC 2016-12-31            NaN   
1021   10051                        HANGER INC 2017-01-31            NaN   
1022   10051                        HANGER INC 2017-02-28            NaN   
1023   10051                        HANGER INC 2017-03-31            NaN   
1024   10051                        HANGER INC 2017-04-30            NaN   
1025   10051                        HANGER INC 2017-05-31            NaN   
1026   10051                        HANGER INC 2017-06-30            NaN   
1027   10051                        HANGER INC 2017-07-31            NaN   
1028   10051                        HANGER INC 2017-08-31            NaN   
1029   10051                        HANGER INC 2017-09-30            NaN   
1030   10051                        HANGER INC 2017-10-31            NaN   
1031   10051                        HANGER INC 2017-11-30            NaN   
1032   10051                        HANGER INC 2017-12-31            NaN   
1033   10051                        HANGER INC 2018-01-31            NaN   
1034   10051                        HANGER INC 2018-09-30            NaN   
1374   10066  FRANKLIN TELECOMMUNICATIONS CORP 1999-03-31            NaN   
1378   10066  FRANKLIN TELECOMMUNICATIONS CORP 1999-07-31            NaN   
1380   10066  FRANKLIN TELECOMMUNICATIONS CORP 1999-09-30            NaN   
1383   10066  FRANKLIN TELECOMMUNICATIONS CORP 1999-12-31            NaN   
2956   10137              ALLEGHENY ENERGY INC 2010-11-30            NaN   
2957   10137              ALLEGHENY ENERGY INC 2010-12-31            NaN   

        price  share_outstanding  
1010      NaN              35291  
1011      NaN              35291  
1012      NaN              35291  
1013      NaN              35291  
1014      NaN              35291  
1015      NaN              35291  
1016      NaN              35291  
1017      NaN              35291  
1018      NaN              35291  
1019      NaN              35291  
1020      NaN              35291  
1021      NaN              35291  
1022      NaN              35291  
1023      NaN              35291  
1024      NaN              35291  
1025      NaN              35291  
1026      NaN              35291  
1027      NaN              35291  
1028      NaN              35291  
1029      NaN              35291  
1030      NaN              35291  
1031      NaN              35291  
1032      NaN              35291  
1033      NaN              35291  
1034  20.8300              36799  
1374   2.9375              24699  
1378      NaN              25207  
1380      NaN              26959  
1383      NaN              28258  
2956  22.8200             169939  
2957      NaN             169939

In [211]:
df.tail()

permno          company_name       date  total_returns  price  \
2955   10137  ALLEGHENY ENERGY INC 2010-10-31      -0.053834  23.20   
2956   10137  ALLEGHENY ENERGY INC 2010-11-30            NaN  22.82   
2957   10137  ALLEGHENY ENERGY INC 2010-12-31            NaN    NaN   
2958   10137  ALLEGHENY ENERGY INC 2011-01-31       0.063531  25.78   
2959   10137  ALLEGHENY ENERGY INC 2011-02-28      -0.009073    NaN   

      share_outstanding  
2955             169615  
2956             169939  
2957             169939  
2958             169939  
2959             169974

In [212]:
# Dropping NaN values (involves dropping the whole row) - we drop the NaNs by writing command dropna()
df.dropna().tail() 

permno          company_name       date  total_returns  price  \
2952   10137  ALLEGHENY ENERGY INC 2010-07-31       0.102514  22.80   
2953   10137  ALLEGHENY ENERGY INC 2010-08-31      -0.010965  22.55   
2954   10137  ALLEGHENY ENERGY INC 2010-09-30       0.094013  24.52   
2955   10137  ALLEGHENY ENERGY INC 2010-10-31      -0.053834  23.20   
2958   10137  ALLEGHENY ENERGY INC 2011-01-31       0.063531  25.78   

      share_outstanding  
2952             169579  
2953             169615  
2954             169615  
2955             169615  
2958             169939

In [213]:
# Method 1 for dealing with all NaN values
df.replace(np.nan,0).tail()

permno          company_name       date  total_returns  price  \
2955   10137  ALLEGHENY ENERGY INC 2010-10-31      -0.053834  23.20   
2956   10137  ALLEGHENY ENERGY INC 2010-11-30       0.000000  22.82   
2957   10137  ALLEGHENY ENERGY INC 2010-12-31       0.000000   0.00   
2958   10137  ALLEGHENY ENERGY INC 2011-01-31       0.063531  25.78   
2959   10137  ALLEGHENY ENERGY INC 2011-02-28      -0.009073   0.00   

      share_outstanding  
2955             169615  
2956             169939  
2957             169939  
2958             169939  
2959             169974

In [216]:
df.head()

permno company_name       date  total_returns    price  share_outstanding
0   10001   ENERGY INC 2009-12-31       0.162621  10.3000               4361
1   10001   ENERGY INC 2010-01-31      -0.018932  10.0600               4361
2   10001   ENERGY INC 2010-02-28      -0.000656  10.0084               4361
3   10001   ENERGY INC 2010-03-31       0.020643  10.1700               4361
4   10001   ENERGY INC 2010-04-30       0.124385  11.3900               6070

In [215]:
# Method 2 for dealing with all NaN values
df['total_returns'] = df['total_returns'].fillna(0)
df['price'] = df['price'].fillna(method='ffill') # ffill is forward fill for handling NaN values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2960 entries, 0 to 2959
Data columns (total 6 columns):
permno               2960 non-null int64
company_name         2960 non-null object
date                 2960 non-null datetime64[ns]
total_returns        2960 non-null float64
price                2960 non-null float64
share_outstanding    2960 non-null int64
dtypes: datetime64[ns](1), float64(2), int64(2), object(1)
memory usage: 138.8+ KB


In [221]:
# Filtering
df = df[(df['permno']!=10001)|(df['permno']!=10006)] # df and index start without 10001 permno
df[df['date'].dt.month==12]

permno                      company_name       date  total_returns  \
218    10014                A J INDUSTRIES INC 1975-12-31      -0.117647   
230    10014                A J INDUSTRIES INC 1976-12-31       0.200000   
244    10028  DALLAS GOLD & SILVER EXCHANGE IN 1992-12-31       0.000000   
256    10028  DALLAS GOLD & SILVER EXCHANGE IN 1993-12-31       0.300000   
268    10028  DALLAS GOLD & SILVER EXCHANGE IN 1994-12-31      -0.052632   
280    10028  DALLAS GOLD & SILVER EXCHANGE IN 1995-12-31       0.111111   
292    10028  DALLAS GOLD & SILVER EXCHANGE IN 1996-12-31      -0.055556   
304    10028  DALLAS GOLD & SILVER EXCHANGE IN 1997-12-31       0.257143   
316    10028  DALLAS GOLD & SILVER EXCHANGE IN 1998-12-31       0.487805   
324    10028             D G S E COMPANIES INC 2007-12-31      -0.018692   
336    10028             D G S E COMPANIES INC 2008-12-31       0.360000   
348    10028             D G S E COMPANIES INC 2009-12-31       0.029197   
360    10028             D G S E COMPANIES INC 2010-12-31       0.043903   
372    10028             D G S E COMPANIES INC 2011-12-31      -0.036129   
377    10028             D G S E COMPANIES INC 2012-12-31      -0.036551   
389    10028             D G S E COMPANIES INC 2013-12-31      -0.074689   
401    10028             D G S E COMPANIES INC 2014-12-31      -0.096296   
413    10028             D G S E COMPANIES INC 2015-12-31      -0.102285   
425    10028             D G S E COMPANIES INC 2016-12-31       0.305263   
437    10028             D G S E COMPANIES INC 2017-12-31      -0.104808   
449    10028             D G S E COMPANIES INC 2018-12-31      -0.043270   
461    10028                       ENVELA CORP 2019-12-31      -0.035714   
462    10029                       A I F S INC 1987-12-31       0.076923   
474    10029                       A I F S INC 1988-12-31      -0.056180   
486    10029                       A I F S INC 1989-12-31       0.583333   
499    10042             CANYON RESOURCES CORP 1996-12-31       0.024390   
511    10042             CANYON RESOURCES CORP 1997-12-31      -0.052632   
523    10042             CANYON RESOURCES CORP 1998-12-31      -0.200000   
535    10042             CANYON RESOURCES CORP 1999-12-31       0.000000   
547    10042             CANYON RESOURCES CORP 2000-12-31       0.000000   
...      ...                               ...        ...            ...   
2609   10137       ALLEGHENY POWER SYSTEMS INC 1981-12-31      -0.012444   
2621   10137       ALLEGHENY POWER SYSTEMS INC 1982-12-31       0.058696   
2633   10137       ALLEGHENY POWER SYSTEMS INC 1983-12-31      -0.025778   
2645   10137       ALLEGHENY POWER SYSTEMS INC 1984-12-31       0.001702   
2657   10137       ALLEGHENY POWER SYSTEMS INC 1985-12-31       0.024242   
2669   10137       ALLEGHENY POWER SYSTEMS INC 1986-12-31      -0.053053   
2681   10137       ALLEGHENY POWER SYSTEMS INC 1987-12-31       0.048443   
2693   10137       ALLEGHENY POWER SYSTEMS INC 1988-12-31       0.003816   
2705   10137       ALLEGHENY POWER SYSTEMS INC 1989-12-31       0.031178   
2717   10137       ALLEGHENY POWER SYSTEMS INC 1990-12-31       0.007759   
2729   10137       ALLEGHENY POWER SYSTEMS INC 1991-12-31       0.012291   
2741   10137       ALLEGHENY POWER SYSTEMS INC 1992-12-31       0.027652   
2753   10137       ALLEGHENY POWER SYSTEMS INC 1993-12-31       0.030048   
2765   10137       ALLEGHENY POWER SYSTEMS INC 1994-12-31       0.018851   
2777   10137       ALLEGHENY POWER SYSTEMS INC 1995-12-31       0.046667   
2789   10137       ALLEGHENY POWER SYSTEMS INC 1996-12-31       0.014156   
2801   10137              ALLEGHENY ENERGY INC 1997-12-31       0.086351   
2813   10137              ALLEGHENY ENERGY INC 1998-12-31       0.033050   
2825   10137              ALLEGHENY ENERGY INC 1999-12-31      -0.056293   
2837   10137              ALLEGHENY ENERGY INC 2000-12-31       0.164491   
2849   10137              ALLEGHENY ENERGY INC 2001-12

In [223]:
# Exporting your dataframe to a csv
df.to_csv('out.csv')

### Merging

In [224]:
# Creating a dataframe with stock names and stock price
df1 = pd.DataFrame([['A',100],
                    ['B',120],
                    ['C',50]],columns=['stock','price'])
df1

stock  price
0     A    100
1     B    120
2     C     50

In [225]:
# Creating a dataframe with stock name and stock PE ratio
df2 = pd.DataFrame([['B',10],
                    ['C',14],
                    ['D',1000]],columns=['stock','pe_ratio'])
df2

stock  pe_ratio
0     B        10
1     C        14
2     D      1000

In [229]:
# Simple merging syntax
pd.merge(left=df1,right=df2,on=['stock'],how='right')
# inner - intersection
# outer - union
# left - keeps all left dataframe 
# right - keeps all right dataframe

stock  price  pe_ratio
0     B  120.0        10
1     C   50.0        14
2     D    NaN      1000

### Groupby

Similar to the Pivot table in excel, this function lets you group/split data for calculations and transformations.

In [230]:
# Reloading the dataset analyzed above (shortcut to parse the dates while reading the data)
df = pd.read_csv("stock_data.csv",sep = ',',parse_dates=['date'])

# Converting any letters into NaN
df['price'] = pd.to_numeric(df['price'],errors='coerce')
df['total_returns'] = pd.to_numeric(df['total_returns'],errors='coerce')

# Creating a new column
df['market cap'] = df['price']*df['share_outstanding']

# Printing the tail
df.tail(7)

permno          company_name       date  total_returns  price  \
2953   10137  ALLEGHENY ENERGY INC 2010-08-31      -0.010965  22.55   
2954   10137  ALLEGHENY ENERGY INC 2010-09-30       0.094013  24.52   
2955   10137  ALLEGHENY ENERGY INC 2010-10-31      -0.053834  23.20   
2956   10137  ALLEGHENY ENERGY INC 2010-11-30            NaN  22.82   
2957   10137  ALLEGHENY ENERGY INC 2010-12-31            NaN    NaN   
2958   10137  ALLEGHENY ENERGY INC 2011-01-31       0.063531  25.78   
2959   10137  ALLEGHENY ENERGY INC 2011-02-28      -0.009073    NaN   

      share_outstanding  market cap  
2953             169615  3824818.25  
2954             169615  4158959.80  
2955             169615  3935068.00  
2956             169939  3878007.98  
2957             169939         NaN  
2958             169939  4381027.42  
2959             169974         NaN

In [235]:
# Grouping by Company Name and counting number of data points available
print(df.groupby('permno')['total_returns'].std())
df[df['permno']==10006]['total_returns'].std()

permno
10001    0.086980
10006    0.088196
10014    0.162146
10028    0.199012
10029    0.140841
10042    0.228634
10048    0.153554
10051    0.155655
10057    0.124718
10064    0.149715
10066    0.285377
10071    0.117875
10085    0.158000
10092    0.162058
10097    0.204756
10102    0.097431
10104    0.051786
10108    0.100021
10116    0.245405
10119    0.091172
10120    0.111890
10123    0.161970
10125    0.326003
10126    0.082606
10137    0.076652
Name: total_returns, dtype: float64


0.08819586127076197

In [ ]:
# Aggregating various statistics of returns for each PERMNO
df.groupby('permno')['total_returns'].agg([np.mean,np.median,np.std,np.var])

In [ ]:
# Grouping by dates and calculating average returns for each date
df.groupby('date')['total_returns'].mean()#.reset_index()

In [ ]:
# Binning market cap >200,000 USD
df.loc[df['market cap']<=200000,'category'] = 1
df.loc[df['market cap']>200000,'category'] = 2
df.head()

In [ ]:
# Grouping by two columns
df.groupby(['date','category'])['total_returns'].mean().head(15)#reset_index

### Stack/Unstack

In [ ]:
# Simple example of unstacking (from above example)
df.groupby(['date','category'])['total_returns'].mean().head(15)
# Execute the unstack command and show how 1 and 2 are now separate columns

In [ ]:
# creating a new dataframe
df2 = pd.DataFrame(np.linspace(0,10,10).reshape(5,2),columns=['A','B'])
df2

In [ ]:
# Stacking the columns 
df2.stack()

# Take Home Questions

- **Pandas Mul Function** - Use this dataframe created in the code chunk below. We have three columns with random float values. The goal is to multiply columns 1 and 2 with columns 3. Firstly, notice how running the function (df[['col1','col2']]*df['col3']) returns a matrix of NaN values. Now, use the .mul() function by passing df[['col1','col2']].mul(df['col3'],axis=0). Notice how we get the same NaN matrix when we change axis=1 instead. 

In [290]:
# Creating a dataframe for problem 1
np.random.seed(123)
df = pd.DataFrame() # Creating a blank data-frame
df['col1'] = np.random.rand(10)
df['col2'] = np.random.rand(10)
df['col3'] = np.random.rand(10)
df[['col1','col2']].mul(df['col3'],axis=0)

col1      col2
0  0.441841  0.217712
1  0.243056  0.619278
2  0.164344  0.317726
3  0.336866  0.036465
4  0.519776  0.287564
5  0.136646  0.238342
6  0.354829  0.066023
7  0.156321  0.040049
8  0.141256  0.156124
9  0.247417  0.335571

- **Load CSV data** (*stock_data.csv*) - parse dates using the *pd.to_datetime* command, read the stock returns as numeric values using the *pd.to_numeric* command, drop all NaN values using the dropna(), and reset the index. We will be working with this dataframe (unless stated otherwise) for the remainder of the questions below.

- **Aggregating** - For each PERMNO, calculate the following statistics for the total_returns - sum, mean, std dev, and median. *Hint - use .groupby('permno') and then aggregate by the stats(.agg() function might be useful)*.

- **Filtering Text** - You can parse the company name column to search for companies that match your key words. A basic filter would be to check for "OIL", "GAS", and "ENERGY" companies. Write a program that returns the number of unique PERMNOs that match the above 3 key-words. *Hint*: Try using df['company_name'].str.contains(). If you are interested in text parsing in Python, **regex** (regular expression) is a powerful library you may want to try out!

- **Merging** - Create a new dataframe with the last available date and price for each PERMNO. For example, the only entry for PERMNO 10001 is 31-Jul-2017 (date) and 12.95 (price). Merge this new dataframe onto the original dataframe. Fill the missing/NaN values using pandas fillna function with a method of your choice.

- **Lagged Market Cap** - Create three new columns that contain the 6-month, 1-year, and 5-year lagged market cap. For example: PERMNO 10001 on 30-Jun-2016 will have the market cap as on 31-Dec-2015 (7.45 $\cdot$ 10505 = 78262.25) as its 6-month lag market cap. Perform a similar exercise to generate the 1-year and the 5-year lag market caps.

- **Resampling Frequency** - Convert the monthly dataframe that you loaded above into quarterly data and annual data. *Hint*: First, load and clean the data as per procedure mentioned above. Second, create a new column containing quarters corresponding to the given month (ex: 30-Nov-2010 is a quarter 4). Lastly, create a new dataframe "df_quarter" by aggregating data using the groupby command. For the quarterly dataframe your columns will be {permno, date, quarter, total_returns, market_cap}. Note that you will need to add the monthly returns to generate quarterly returns. Also note that you will need the quarter end market cap for each PERMNO each quarter. For example : In the Quarterly dataframe, PERMNO 10001 as of 31-Mar-2010 will have total returns 0.001055 (-0.018932-0.000656+0.020643) and the market cap 44351.37 (10.17$\cdot$4361). Perform the same procedure for creating data at an annual frequency.

In [ ]:
# Question 1
np.random.seed(123)
df = pd.DataFrame() # Creating a blank data-frame
df['col1'] = np.random.rand(10)
df['col2'] = np.random.rand(10)
df['col3'] = np.random.rand(10)
df[['col1','col2']].mul(df['col3'],axis=0)

In [298]:
# Question 2
df = pd.read_csv("stock_data.csv",sep = ',',parse_dates=['date'])
df['total_returns'] = pd.to_numeric(df['total_returns'],errors='coerce')
df['price'] = pd.to_numeric(df['price'],errors='coerce')
print(df.shape)
df = df.dropna().reset_index(drop=True)
print(df.shape)
df.head()

(2960, 6)
(2913, 6)


permno company_name       date  total_returns    price  share_outstanding
0   10001   ENERGY INC 2009-12-31       0.162621  10.3000               4361
1   10001   ENERGY INC 2010-01-31      -0.018932  10.0600               4361
2   10001   ENERGY INC 2010-02-28      -0.000656  10.0084               4361
3   10001   ENERGY INC 2010-03-31       0.020643  10.1700               4361
4   10001   ENERGY INC 2010-04-30       0.124385  11.3900               6070

In [292]:
# Question 3
energy = df[df['company_name'].str.contains("ENERGY")]['permno'].nunique()
oil = df[df['company_name'].str.contains("OIL")]['permno'].nunique()
gas = df[df['company_name'].str.contains("GAS")]['permno'].nunique()

print("There are {} energy and {} oil & gas firms".format(energy,oil+gas))

There are 2 energy and 1 oil & gas firms


In [299]:
# Question 4
# First selecting each PERMNO and selecting the last avaialbale date
df_last = df.groupby('permno')['date'].max().reset_index()

# Merging the corresponding price of the last date
df_last = df_last.merge(df[['permno','date','price']],on=['permno','date'],how='left')

# Printing the output of the above dataframe
df_last = df_last.rename(columns={'price':'price_last'}) #renaming price to price_last
print(df_last.head())

# Merging df_last onto our original dataframe
df = df.merge(df_last,on=['permno','date'],how='left')

# Filling NaN using bfill method
df['price_last'] = df['price_last'].fillna(method='bfill')
df.head()

   permno       date  price_last
0   10001 2017-07-31      12.950
1   10006 1984-05-31      52.625
2   10014 1977-03-31       4.750
3   10028 2019-12-31       1.350
4   10029 1990-07-31       6.125


permno company_name       date  total_returns    price  share_outstanding  \
0   10001   ENERGY INC 2009-12-31       0.162621  10.3000               4361   
1   10001   ENERGY INC 2010-01-31      -0.018932  10.0600               4361   
2   10001   ENERGY INC 2010-02-28      -0.000656  10.0084               4361   
3   10001   ENERGY INC 2010-03-31       0.020643  10.1700               4361   
4   10001   ENERGY INC 2010-04-30       0.124385  11.3900               6070   

   price_last  
0       12.95  
1       12.95  
2       12.95  
3       12.95  
4       12.95

In [300]:
# First selecting each PERMNO and selecting the last avaialbale date
df_last = df.groupby('permno')['date'].max().reset_index()

# Merging the corresponding price of the last date
df_last = df_last.merge(df[['permno','date','price']],on=['permno','date'],how='left')

df = df.merge(df_last,on=['permno','date'],how='left')

df[df['permno']==10001]

permno     company_name       date  total_returns  price_x  \
0    10001       ENERGY INC 2009-12-31       0.162621  10.3000   
1    10001       ENERGY INC 2010-01-31      -0.018932  10.0600   
2    10001       ENERGY INC 2010-02-28      -0.000656  10.0084   
3    10001       ENERGY INC 2010-03-31       0.020643  10.1700   
4    10001       ENERGY INC 2010-04-30       0.124385  11.3900   
5    10001       ENERGY INC 2010-05-31       0.004829  11.4000   
6    10001       ENERGY INC 2010-06-30      -0.043421  10.8600   
7    10001  GAS NATURAL INC 2010-07-31       0.083333  11.7200   
8    10001  GAS NATURAL INC 2010-08-31      -0.111263  10.3710   
9    10001  GAS NATURAL INC 2010-09-30       0.076560  11.1200   
10   10001  GAS NATURAL INC 2010-10-31       0.032815  11.4399   
11   10001  GAS NATURAL INC 2010-11-30      -0.118436  10.0400   
12   10001  GAS NATURAL INC 2010-12-31       0.052291  10.5200   
13   10001  GAS NATURAL INC 2011-01-31       0.028992  10.7800   
14   10001  GAS NATURAL INC 2011-02-28       0.022727  10.9800   
15   10001  GAS NATURAL INC 2011-03-31       0.072404  11.7300   
16   10001  GAS NATURAL INC 2011-04-30      -0.038789  11.2300   
17   10001  GAS NATURAL INC 2011-05-31       0.028050  11.5000   
18   10001  GAS NATURAL INC 2011-06-30       0.008261  11.5500   
19   10001  GAS NATURAL INC 2011-07-31      -0.028139  11.1800   
20   10001  GAS NATURAL INC 2011-08-31      -0.013864  10.9800   
21   10001  GAS NATURAL INC 2011-09-30       0.005009  10.9900   
22   10001  GAS NATURAL INC 2011-10-31       0.005005  11.0000   
23   10001  GAS NATURAL INC 2011-11-30      -0.005000  10.9000   
24   10001  GAS NATURAL INC 2011-12-31       0.051835  11.4200   
25   10001  GAS NATURAL INC 2012-01-31      -0.019702  11.1500   
26   10001  GAS NATURAL INC 2012-02-29       0.005005  11.1608   
27   10001  GAS NATURAL INC 2012-03-31       0.048760  11.6600   
28   10001  GAS NATURAL INC 2012-04-30      -0.015000  11.4401   
29   10001  GAS NATURAL INC 2012-05-31      -0.084361  10.4300   
..     ...              ...        ...            ...      ...   
62   10001  GAS NATURAL INC 2015-02-28       0.035897  10.1000   
63   10001  GAS NATURAL INC 2015-03-31      -0.000495   9.9600   
64   10001  GAS NATURAL INC 2015-04-30       0.011044  10.0700   
65   10001  GAS NATURAL INC 2015-05-31      -0.006951  10.0000   
66   10001  GAS NATURAL INC 2015-06-30       0.030000  10.3000   
67   10001  GAS NATURAL INC 2015-07-31      -0.013107  10.0300   
68   10001  GAS NATURAL INC 2015-08-31      -0.111665   8.9100   
69   10001  GAS NATURAL INC 2015-09-30       0.003367   8.9400   
70   10001  GAS NATURAL INC 2015-10-31       0.001678   8.8200   
71   10001  GAS NATURAL INC 2015-11-30      -0.009070   8.7400   
72   10001  GAS NATURAL INC 2015-12-31      -0.132151   7.4500   
73   10001  GAS NATURAL INC 2016-01-31       0.116779   8.3200   
74   10001  GAS NATURAL INC 2016-02-29      -0.055288   7.8600   
75   10001  GAS NATURAL INC 2016-03-31      -0.006361   7.8100   
76   10001  GAS NATURAL INC 2016-04-30      -0.055698   7.3000   
77   10001  GAS NATURAL INC 2016-05-31      -0.021918   7.1400   
78   10001  GAS NATURAL INC 2016-06-30      -0.021008   6.9900   
79   10001  GAS NATURAL INC 2016-07-31       0.026466   7.1000   
80   10001  GAS NATURAL INC 2016-08-31       0.036620   7.3600   
81   10001  GAS NATURAL INC 2016-09-30       0.042120   7.6700   
82   10001  GAS NATURAL INC 2016-10-31       0.619948  12.3500   
83   10001  GAS NATURAL INC 2016-11-30       0.012146  12.5000   
84   10001  GAS NATURAL INC 2016-12-31       0.010000  12.5500   
85   10001  GAS NATURAL INC 2017-01-31       0.007968  12.6500   
86   10001  GAS NATURAL INC 2017-02-28       0.000000  12.6500   
87   10001  GAS NATURAL INC 2017-03-31       0.009881  12.7000   
88   10001  GAS NATURAL INC 2017-04-30      -0.015748  12.5000   
89   10001  GAS NATURAL INC 2017-05-31       0.016000  12.7000   
90   10001  GAS NATURAL INC 2017-06-30      

In [302]:
# Question 5
# Loading and cleaning the dataframe
df = pd.read_csv("stock_data.csv",sep = ',',parse_dates=['date'])
df['total_returns'] = pd.to_numeric(df['total_returns'],errors='coerce')
df['price'] = pd.to_numeric(df['price'],errors='coerce')
df = df.dropna().reset_index(drop=True) # Passing multiple functions (big plus of Python!)

# Computing market cap
df['market_cap'] = df['price']*df['share_outstanding']

# Creating 6-month lag column for each date
df['date_lag_6m'] = df['date']- pd.tseries.offsets.MonthEnd(6)

# Creating 1-year and 5-year lag columns for each date
df['date_lag_1y'] = df['date']- pd.tseries.offsets.MonthEnd(12)
df['date_lag_5y'] = df['date']- pd.tseries.offsets.MonthEnd(60)

# Merging market cap for lag 6-months column
df = df.merge(df[['permno',
                  'date',
                  'market_cap']].rename(columns={'date':'date_lag_6m',
                                                 'market_cap':'market_cap_lag_6m'}),on=['permno','date_lag_6m'],how='left')

df.head(10)

# Note : You don't want to apply the shift operator since firms could get listed/delisted/listed again
# Similarly you can do this for 1Y and 5Y lags

permno     company_name       date  total_returns    price  \
0   10001       ENERGY INC 2009-12-31       0.162621  10.3000   
1   10001       ENERGY INC 2010-01-31      -0.018932  10.0600   
2   10001       ENERGY INC 2010-02-28      -0.000656  10.0084   
3   10001       ENERGY INC 2010-03-31       0.020643  10.1700   
4   10001       ENERGY INC 2010-04-30       0.124385  11.3900   
5   10001       ENERGY INC 2010-05-31       0.004829  11.4000   
6   10001       ENERGY INC 2010-06-30      -0.043421  10.8600   
7   10001  GAS NATURAL INC 2010-07-31       0.083333  11.7200   
8   10001  GAS NATURAL INC 2010-08-31      -0.111263  10.3710   
9   10001  GAS NATURAL INC 2010-09-30       0.076560  11.1200   

   share_outstanding  market_cap date_lag_6m date_lag_1y date_lag_5y  \
0               4361  44918.3000  2009-06-30  2008-12-31  2004-12-31   
1               4361  43871.6600  2009-07-31  2009-01-31  2005-01-31   
2               4361  43646.6324  2009-08-31  2009-02-28  2005-02-28   
3               4361  44351.3700  2009-09-30  2009-03-31  2005-03-31   
4               6070  69137.3000  2009-10-31  2009-04-30  2005-04-30   
5               6071  69209.4000  2009-11-30  2009-05-31  2005-05-31   
6               6080  66028.8000  2009-12-31  2009-06-30  2005-06-30   
7               6080  71257.6000  2010-01-31  2009-07-31  2005-07-31   
8               6073  62983.0830  2010-02-28  2009-08-31  2005-08-31   
9               6073  67531.7600  2010-03-31  2009-09-30  2005-09-30   

   market_cap_lag_6m  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
5                NaN  
6         44918.3000  
7         43871.6600  
8         43646.6324  
9         44351.3700

In [306]:
# Question 6
# Loading and cleaning the dataframe
df = pd.read_csv("stock_data.csv",sep = ',',parse_dates=['date'])
df['total_returns'] = pd.to_numeric(df['total_returns'],errors='coerce')
df['price'] = pd.to_numeric(df['price'],errors='coerce')
df = df.dropna().reset_index(drop=True) # Passing multiple functions (big plus of Python!)

# Computing market cap
df['market_cap'] = df['price']*df['share_outstanding']

# Creating quarters column
df['quarter'] = df['date'].dt.quarter

# Creating year column
df['year'] = df['date'].dt.year

# Selecting quarter end month
df = df.merge(df.groupby(['permno',
                          'year',
                          'quarter'])['date'].max().reset_index().rename(columns={'date':'date_quarter_max'}),on=['permno',
                                                                                                                  'year',
                                                                                                                  'quarter'],how='left')

# Creating df_quarter dataframe with averaged monthly returns for each quarter
df_quarter = df.groupby(['permno','year','quarter','date_quarter_max'])['total_returns'].sum().reset_index()

# Merging quarter end market cap values
df_quarter = df_quarter.merge(df[['date','market_cap']],left_on=['date_quarter_max'],right_on=['date'],how='left')

df_quarter.head()

permno  year  quarter date_quarter_max  total_returns       date  \
0   10001  2009        4       2009-12-31       0.162621 2009-12-31   
1   10001  2009        4       2009-12-31       0.162621 2009-12-31   
2   10001  2009        4       2009-12-31       0.162621 2009-12-31   
3   10001  2009        4       2009-12-31       0.162621 2009-12-31   
4   10001  2009        4       2009-12-31       0.162621 2009-12-31   

   market_cap  
0    44918.30  
1    13865.94  
2   438134.40  
3    21229.68  
4  3981339.24

In [ ]:
df.head(10)